lib/matplotlib/axes/_axes.py

```diff
@@ -6142,6 +6147,16 @@ def _normalize_input(inp, ename='input'):
                 raise ValueError(
                     'weights should have the same shape as x')
 
+        # Combine the masks from x[i] and w[i] (if applicable) into a single
+        # mask and apply it to both.
+        if not input_empty:
+            for i in range(len(x)):
+                mask_i = x[i].mask
+                if w[i] is not None:
+                    mask_i = mask_i | w[i].mask
```

Comment:
Not quite what I had in mind...
```python
xclean, wclean = [], []
for xi, wi in zip(x, w):
    xiclean, wiclean = cbook.delete_masked_points(xi, wi)
    xclean.append(xiclean)
    wclean.append(wiclean)
```
This works regardless of whether wi is None or a sequence the length of xi. Using `cbook.delete_masked_points` is only worthwhile if you are giving it more than one argument.  Deleting the masked points from a single 1-D masked array can be done simply by calling it's `compressed()` method, so that's all you need in your version above.  You would call `w[i] = w[i].compressed()` inside the inner `if`, and `x[i]` = `x[i].compressed()` immediately after it.  There are many variations on the theme; I'm not sure which is best.